pkg/ottl/functions.go

```diff
@@ -18,11 +18,38 @@ import (
 	"errors"
 	"fmt"
 	"reflect"
+	"strconv"
 	"strings"
 
+	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 )
 
+// Arguments holds the arguments for an OTTL function, with arguments
+// specified as fields on a struct. Argument ordering is defined
+type Arguments interface{}
+
+// FunctionContext contains data provided by the Collector
+// component to the OTTL for use in functions.
+type FunctionContext struct {
+	Set component.TelemetrySettings
+}
+
+// Factory defines an OTTL function factory that will generate an OTTL
+// function to be called based on an invocation within a statement.
+type Factory[K any] interface {
```

Comment:
I just realized that this interface is going to allow us to do Optional Parameters in the future 🚀

Reply:
Oh wow, you're right. That's a surprise benefit.

Reply:
I miss something here.

Reply:
@bogdandrutu with the new pattern the CreateFunction function will be able to identify when fields on the Arguments struct aren't set and provide default values to the business logic function.  At least thats how I foresee it at the moment.

Reply:
Unfortunately I'm not sure that would work since it wouldn't be easy to check if e.g. an int struct field is unset or if it is `0`. I think we will have to continue to do a check in the parser for parity between the arguments passed in through the DSL and what arguments the function defines.

I can think of two ways we could use this to provide the ability for functions to have optional parameters:
1. Add an "optional" tag to `Arguments` struct fields that are optional. The parser can then read this tag to determine when a field is optional.
2. Provide an optional `OptionalParameters` interface that factories can implement that provides an index after which all parameters are optional.

I strongly prefer option 1 between these two. There also might be another way to accomplish this that I've missed.

Reply:
I like the sound of option 1 as well, although I think the factory implementor will need a way to provide the default value since it wont always be the same as the type's default value.  We can talk more about Optional Parameters in a dedicated issue.

Reply:
I opened an issue to discuss this here: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20879.